File: leetcode/0600-non-negative-integers-without-consecutive-ones/solution.go
func findIntegers(n int) int {
    f := [31]int{1,2} // f(i) = how many way to pick valid number not larger than 2^i
    for i:=2;i<31;i++ {
        f[i] = f[i-1] + f[i-2]
    }
    ret := 1
    //fmt.Println(f);
    for i:=30;i>=0;i-- {
        pattern := (n>>i) & 0b11
        if pattern == 0b11 {
            // can't pick 1 here, go to the next candidate not larger than n
            // 11xxxxxxxxx -> 00111111111
            n = 1<<i - 1
        } else if pattern == 0b01 {
            // can pick 1 here
            ret += f[i]
        }
    }
    return ret
}

File: leetcode/0600-non-negative-integers-without-consecutive-ones/solution.rs
impl Solution {
    pub fn find_integers(mut n: i32) -> i32 {
        let mut f = vec![1,2]; // f(i) = how many way to pick valid number not larger than 2^i
        for i in 2..32 {
            f.push(f[i-1] + f[i-2]);
        }
        //println!("{f:?}");
        let mut ret = 1;
        for i in (0..32).rev() {
            let pattern = (n>>i) & 0b11;
            match pattern {
                0b11 => {
                    // can't pick 1 here, go to the next candidate not larger than n
                    // 11xxxxxxxxx -> 00111111111
                    n = (1<<i) - 1;
                }
                0b01 => {
                    // can pick 1 here
                    ret += f[i];
                }
                _ => {}
            }
        }
        return ret;
    }
}
